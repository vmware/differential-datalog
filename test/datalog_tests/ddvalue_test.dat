# Add one record per transaction to guarantee deterministic output order
# (the ordering of `Any` may vary across Rust compiler releases).

start;
insert ddvalue_test::DDValTest1("foo", 1, ["bar", "buzz"]),
commit dump_changes;

start;
insert ddvalue_test::DDValTest2([(1, "foo"), (2, "bar")], (true, [5,4,3,2,1])),
commit dump_changes;
