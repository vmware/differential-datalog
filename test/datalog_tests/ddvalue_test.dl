/* Test `Any` API. */

input relation DDValTest1(x: string, y: usize, z: Vec<istring>)
input relation DDValTest2(a: Map<usize, string>, b: (bool, Set<u128>))

output relation DDValTestOutput(relname: string, ddval: Any)

DDValTestOutput("DDValTest1", x.to_any()) :- DDValTest1[x].
DDValTestOutput("DDValTest2", x.to_any()) :- DDValTest2[x].

output relation DDValTest1Output[DDValTest1]
output relation DDValTest2Output[DDValTest2]

DDValTest1Output[from_any(x).unwrap_or_default()] :- DDValTestOutput("DDValTest1", x).
DDValTest2Output[from_any(x).unwrap_or_default()] :- DDValTestOutput("DDValTest2", x).
