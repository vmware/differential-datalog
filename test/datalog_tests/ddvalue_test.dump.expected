ddvalue_test::DDValTest1Output:
ddvalue_test::DDValTest1{.x = "foo", .y = 1, .z = ["bar", "buzz"]}: +1
ddvalue_test::DDValTestOutput:
ddvalue_test::DDValTestOutput{.relname = "DDValTest1", .ddval = ddvalue_test::DDValTest1{.x = "foo", .y = 1, .z = ["bar", "buzz"]}}: +1
ddvalue_test::DDValTest2Output:
ddvalue_test::DDValTest2{.a = [(1, "foo"), (2, "bar")], .b = (true, [1, 2, 3, 4, 5])}: +1
ddvalue_test::DDValTestOutput:
ddvalue_test::DDValTestOutput{.relname = "DDValTest2", .ddval = ddvalue_test::DDValTest2{.a = [(1, "foo"), (2, "bar")], .b = (true, [1, 2, 3, 4, 5])}}: +1
